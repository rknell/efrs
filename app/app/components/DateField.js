import React, { Component } from 'react'
import { View } from 'react-native'
import { observer } from 'mobx-react/native'
import { computed } from 'mobx'
import EditModeWrapper from './EditModeWrapper'
import moment from 'moment'
import DateFieldView from '../componentViews/DateFieldView'
import FieldController from '../services/FieldController'
import GlobalStyle from '../services/GlobalStyle'

@observer
class DateField extends Component {
  constructor (props) {
    super(props)
    this.controller = new DateFieldController(props)
  }

  componentWillMount () {
    this.controller.initialise()
    if (!this.props.data.values.default && this.props.editMode !== true) {
      if (this.controller.settingsObject['Default Date Offset (Days)']) {
        this.props.data.values.default = moment().add(this.controller.settingsObject['Default Date Offset (Days)'].value, 'days').toDate()
      } else if (this.controller.settingsObject['Default to Today'] && this.props.isSearching !== true) {
        this.props.data.values.default = new Date()
      }
    }
  }

  render () {
    if (this.props.isSearching) return null
    return (
      <View>
        <EditModeWrapper
          {...this.props}
          onPressEdit={() => {
            this.controller.viewSettings()
          }}>
          <DateFieldView
            globalStyle={GlobalStyle.style}
            maxDate={this.controller.maxDate}
            minDate={this.controller.minDate}
            onChange={(date) => {
              this.props.data.values.default = date
            }}
            value={this.props.data.values.default}
            settings={this.controller.settingsObject}
            editMode={this.props.editMode}
            readOnly={this.props.readOnly}
          />
        </EditModeWrapper>
      </View>
    )
  }
}

class DateFieldController extends FieldController {
  constructor (props) {
    super(props)
    this.props = props
  }

  @computed
  get maxDate () {
    if (this.settingsObject['Max Days Future'] !== '') {
      return moment().add(Number(this.settingsObject['Max Days Future']), 'days').toDate()
    } else {
      return null
    }
  }

  @computed
  get minDate () {
    if (this.settingsObject['Max Days Past'] !== '') {
      return moment().subtract(Number(this.settingsObject['Max Days Past']), 'days').toDate()
    } else {
      return null
    }
  }

  settings = [{
    value: 'Date',
    label: 'Label',
    fieldType: 'TextFieldView',
    extraProps: {}
  }, {
    value: false,
    label: 'Show In Search',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: '',
    label: 'Max Days Future',
    fieldType: 'TextFieldView'
  }, {
    value: '',
    label: 'Max Days Past',
    fieldType: 'TextFieldView'
  }, {
    value: '',
    label: 'Default Date Offset (Days)',
    fieldType: 'TextFieldView'
  }, {
    value: false,
    label: 'Default to Today',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: false,
    label: 'Required',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: false,
    label: 'Read Only',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: false,
    label: 'Hidden',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: false,
    label: 'Report When Hidden',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }, {
    value: 'date',
    label: 'Data Type',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Date', value: 'date' }, { label: 'Time', value: 'time' }, { label: 'Date & Time', value: 'datetime' }]
    }
  }, {
    value: 'calendar',
    label: 'Initial Picker',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Year', value: 'year' }, { label: 'Calendar', value: 'calendar' }]
    }
  }, {
    value: false,
    label: 'Set Now Button',
    fieldType: 'QuickPickView',
    extraProps: {
      options: [{ label: 'Yes', value: true }, { label: 'No', value: false }]
    }
  }]
}

export default DateField
