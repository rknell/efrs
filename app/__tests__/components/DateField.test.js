import React from 'react'
import DateField from '../../app/components/DateField'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

it('should match the snapshot', () => {
  const wrapper = shallow(<DateField data={{
    settings: [],
    values: {
      default: ''
    }
  }} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})
