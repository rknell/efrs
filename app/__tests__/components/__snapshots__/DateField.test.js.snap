// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<View>
  <EditModeWrapper
    data={
      Object {
        "settings": Array [
          Object {
            "calculated": null,
            "extraProps": Object {},
            "fieldType": "TextFieldView",
            "label": "Label",
            "value": "Date",
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Yes",
                  "value": true,
                },
                Object {
                  "label": "No",
                  "value": false,
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Show In Search",
            "value": false,
          },
          Object {
            "calculated": null,
            "fieldType": "TextFieldView",
            "label": "Max Days Future",
            "value": "",
          },
          Object {
            "calculated": null,
            "fieldType": "TextFieldView",
            "label": "Max Days Past",
            "value": "",
          },
          Object {
            "calculated": null,
            "fieldType": "TextFieldView",
            "label": "Default Date Offset (Days)",
            "value": "",
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Yes",
                  "value": true,
                },
                Object {
                  "label": "No",
                  "value": false,
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Default to Today",
            "value": false,
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Yes",
                  "value": true,
                },
                Object {
                  "label": "No",
                  "value": false,
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Required",
            "value": false,
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Yes",
                  "value": true,
                },
                Object {
                  "label": "No",
                  "value": false,
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Read Only",
            "value": false,
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Yes",
                  "value": true,
                },
                Object {
                  "label": "No",
                  "value": false,
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Hidden",
            "value": false,
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Yes",
                  "value": true,
                },
                Object {
                  "label": "No",
                  "value": false,
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Report When Hidden",
            "value": false,
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Date",
                  "value": "date",
                },
                Object {
                  "label": "Time",
                  "value": "time",
                },
                Object {
                  "label": "Date & Time",
                  "value": "datetime",
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Data Type",
            "value": "date",
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Year",
                  "value": "year",
                },
                Object {
                  "label": "Calendar",
                  "value": "calendar",
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Initial Picker",
            "value": "calendar",
          },
          Object {
            "calculated": null,
            "extraProps": Object {
              "options": Array [
                Object {
                  "label": "Yes",
                  "value": true,
                },
                Object {
                  "label": "No",
                  "value": false,
                },
              ],
            },
            "fieldType": "QuickPickView",
            "label": "Set Now Button",
            "value": false,
          },
        ],
        "values": Object {
          "default": "",
        },
      }
    }
    onPressEdit={[Function]}
  >
    <DateFieldView
      globalStyle={
        Object {
          "disabledText": "rgb(179, 179, 179)",
          "errorColour": "red",
          "fullLogo": Object {
            "testUri": "../../../app/assets/efrsFullBanner.png",
          },
          "neutralColour": "white",
          "neutralColourDisabled": "rgba(255, 255, 255, 0.3)",
          "neutralColourHighlight": "rgb(217, 217, 217)",
          "neutralColourIsDark": false,
          "neutralColourOffset": "rgb(242, 242, 242)",
          "neutralColourText": "rgb(51, 51, 51)",
          "neutralColourTextDisabled": "rgba(51, 51, 51, 0.3)",
          "primaryColour": "dodgerblue",
          "primaryColourDisabled": "rgba(30, 144, 255, 0.3)",
          "primaryColourHighlight": "rgb(106, 182, 255)",
          "primaryColourIsDark": true,
          "primaryColourText": "rgb(255, 255, 255)",
          "primaryColourTextDisabled": "rgba(255, 255, 255, 0.3)",
          "secondaryColour": "#f3f3f3",
          "secondaryColourDisabled": "rgba(243, 243, 243, 0.3)",
          "secondaryColourHighlight": "rgb(205, 205, 205)",
          "secondaryColourIsDark": undefined,
          "secondaryColourText": "rgb(39, 39, 39)",
          "secondaryColourTextDisabled": "rgba(39, 39, 39, 0.3)",
          "secondaryLogo": Object {
            "testUri": "../../../app/assets/icon.png",
          },
        }
      }
      maxDate={null}
      minDate={null}
      onChange={[Function]}
      settings={Object {}}
      value=""
    />
  </EditModeWrapper>
</View>
`;
